actix/src/sec-4-context.md

```diff
@@ -1 +1,143 @@
-[WIP] 
\ No newline at end of file
+# Context
+
+Actors all maintain an internal execution context, or state. This
+allows an actor to determine it's own Address, change mailbox limits,
+or stop it's execution.
+
+## Mailbox
+
+All messages go to the actor's mailbox first, then the actor's execution context
+calls specific message handlers. Mailboxes in general are bounded. The capacity is
+specific to the context implementation. For the `Context`  type the capacity is set to
+16 messages by default and can be increased with
+[*Context::set_mailbox_capacity()*](../actix/struct.Context.html#method.set_mailbox_capacity).
+
+```rust
+# extern crate actix;
+# use actix::prelude::*;
+struct MyActor;
+impl Actor for MyActor {
+    type Context = Context<Self>;
+
+    fn started(&mut self, ctx: &mut Self::Context) {
+        ctx.set_mailbox_capacity(1);
+    }
+}
+
+# fn main() {
+# System::new("test");
+let addr = MyActor.start();
+# }
+```
+
+Remember that this doesn't apply to `Addr::do_send(M)`, `AsyncContext::notify(M)` or
+`AsyncContext::notify_later(M, Duration)` which bypass the Mailbox queue limit.
```

Comment:
Ok, but let's be accurate here.

`do_send` bypass Mailbox queue limit

But `notfy` and `notify_later` bypass Mailbox entirely(i.e. message is not queued in Mailbox and immediately gets future that handle the message spawned on Actor's context)
Which means that it is better to use it performance wise.

Reply:
Interesting. So the statement still is "true" that it does bypass the queue limit, but perhaps instead of over complicating this section here with a discussion about the spawning of the future, maybe leave this as "they bypass the mailbox queue" and then we dedicate a proper section on context to notify and notify_later with examples and explanation of their behaviours. I think that as someone who would read this book, I would prefer that layout because when I'm looking at the mailbox section I only care about what does and doesn't interact with it, and then I can learn more about the details of notify in a seperate part.

Is that a reasonable approach you think?

Reply:
Well there is no need to overcomplicate it, you can put it simply as:

Remember that this doesn't apply to `Addr::do_send(M)`, which bypass the Mailbox queue limit, or  `AsyncContext::notify(M)` with `AsyncContext::notify_later(M, Duration)` which bypass the Mailbox entirely

it is important to have docs to distinguish between these too, so that it wouldn't come out as surprise later on

Reply:
Okay, I've made that change. I was reading the code to understand this better and in contextimpl, I notice that the order is "mailbox.poll()" followed by item processing, so would this mean that notify/notify_later events actually are processed once the mailbox queue is cleared? (Also mailbox.poll seems to have no upper bound/backoff on processing, so couldn't it happen that under high mailbox pressure that it never yields to allow notify/notify_later processing? )

Reply:
Friendly reminder that I have updated this documentation for you :)